Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.cpp

```diff
@@ -0,0 +1,320 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "PlatformRawAudioDataCocoa.h"
+
+#if ENABLE(WEB_CODECS) && USE(AVFOUNDATION)
+
+#include "AudioSampleFormat.h"
+#include "CAAudioStreamDescription.h"
+#include "CMUtilities.h"
+#include "MediaSampleAVFObjC.h"
+#include "NotImplemented.h"
+#include "SharedBuffer.h"
+#include "WebAudioBufferList.h"
+#include "WebCodecsAudioDataAlgorithms.h"
+
+#include <CoreAudio/CoreAudioTypes.h>
+#include <wtf/MediaTime.h>
+
+#include <pal/cf/CoreMediaSoftLink.h>
+
+using namespace WTF;
+
+namespace WebCore {
+
+Ref<PlatformRawAudioData> PlatformRawAudioData::create(Ref<MediaSample>&& sample)
+{
+    return PlatformRawAudioDataCocoa::create(downcast<MediaSampleAVFObjC>(WTFMove(sample)));
+}
+
+static AudioStreamDescription::PCMFormat audioSampleFormatToPCMFormat(AudioSampleFormat format)
+{
+    switch (format) {
+    case AudioSampleFormat::U8:
+        return AudioStreamDescription::Uint8;
+    case AudioSampleFormat::S16:
+        return AudioStreamDescription::Int16;
+    case AudioSampleFormat::S32:
+        return AudioStreamDescription::Int32;
+    case AudioSampleFormat::F32:
+        return AudioStreamDescription::Float32;
+    case AudioSampleFormat::U8Planar:
+        return AudioStreamDescription::Uint8;
+    case AudioSampleFormat::S16Planar:
+        return AudioStreamDescription::Int16;
+    case AudioSampleFormat::S32Planar:
+        return AudioStreamDescription::Int32;
+    case AudioSampleFormat::F32Planar:
+        return AudioStreamDescription::Float32;
+    }
+}
+
+static CAAudioStreamDescription::IsInterleaved interleavedFormat(AudioSampleFormat format)
+{
+    return isAudioSampleFormatInterleaved(format) ? CAAudioStreamDescription::IsInterleaved::Yes : CAAudioStreamDescription::IsInterleaved::No;
+}
+
+static RetainPtr<CMSampleBufferRef> createSampleBuffer(const CAAudioStreamDescription& description, const CMTime& time, size_t numberOfFrames, WebAudioBufferList* list)
```

Comment:
```suggestion
static RetainPtr<CMSampleBufferRef> createSampleBuffer(const CAAudioStreamDescription& description, const CMTime& time, size_t numberOfFrames, const WebAudioBufferList* list)
```
?

Reply:
`const` here doesn't make the list const itself; just the pointer. But I see what you mean. If we had the object const itself, The CM method would then need a const_cast which is far from great